pallets/xcm-core-buyer/src/lib.rs

```diff
@@ -0,0 +1,435 @@
+// Copyright (C) Moondance Labs Ltd.
+// This file is part of Tanssi.
+
+// Tanssi is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Tanssi is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Tanssi.  If not, see <http://www.gnu.org/licenses/>
+
+//! # XCM Core Buyer Pallet
+//!
+//! This pallet allows collators to buy parathread cores on demand.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(any(test, feature = "runtime-benchmarks"))]
+mod benchmarks;
+pub mod weights;
+
+use {
+    crate::weights::WeightInfo,
+    dp_core::ParaId,
+    frame_support::{
+        pallet_prelude::*,
+        traits::fungible::{Balanced, Inspect},
+    },
+    frame_system::pallet_prelude::*,
+    sp_runtime::traits::{Convert, Get},
+    sp_std::vec,
+    sp_std::vec::Vec,
+    staging_xcm::prelude::*,
+    staging_xcm::v3::{InteriorMultiLocation, MultiAsset, MultiAssets, Xcm},
+    tp_traits::ParathreadParams,
+};
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+
+    /// Data preservers pallet.
+    #[pallet::pallet]
+    #[pallet::without_storage_info]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// Overarching event type.
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+        type Currency: Inspect<Self::AccountId> + Balanced<Self::AccountId>;
+
+        type XcmSender: SendXcm;
+        type GetPurchaseCoreCall: GetPurchaseCoreCall;
+        type GetBlockNumber: Get<u32>;
+        type GetParathreadAccountId: Convert<ParaId, [u8; 32]>;
+        type SelfParaId: Get<ParaId>;
+        type MaxParathreads: Get<u32>;
+        type GetParathreadParams: GetParathreadParams;
+        type GetAssignedCollators: GetParathreadCollators<Self::AccountId>;
```

Comment:
can we add a comment mentioning why we need this? also the upper variables

Reply:
Documented most config types